add porting guide for bare vars in conditionals (#60943)
ansible/ansible: docs/docsite/rst/porting_guides/porting_guide_2.8.rst

For dictionaries and lists, use the ``length`` filter to evaluate the presence of a dictionary or list as ``True``:

.. code-block:: yaml

      - debug:
        when: my_list | length > 0

      - debug:
        when: my_dictionary | length > 0

Do not use the ``bool`` filter with lists or dictionaries. If you use ``bool`` with a list or dict, Ansible will always evaluate it as ``False``.

Double-interpolation
^^^^^^^^^^^^^^^^^^^^

The ``conditional_bare_variables`` setting also affects variables set based on other variables. The old behavior unexpectedly double-interpolated those variables. For example:

.. code-block:: yaml

    vars:
      double_interpolated: 'bare_variable'
      bare_variable: false

    tasks:
      - debug:
        when: double_interpolated

* In Ansible 2.7 and earlier, ``when: double_interpolated`` evaluated to the value of ``bare_variable``, in this case, ``False``. If the variable ``bare_variable`` is undefined, the conditional fails.
* In Ansible 2.8 and later, with bare variables disabled, Ansible evaluates ``double_interpolated`` as the string ``'bare_variable'``, which is ``True``.

To double-interpolate variable values, use curly braces:

> .. code-block:: yaml

Reviewer: ```suggestion
.. code-block:: yaml+jinja
```

    vars:
      double_interpolated: "{{ other_variable }}"
      other_variable: false

Nested variables
^^^^^^^^^^^^^^^^

> The ``conditional_bare_variables`` setting does not affect nested variables. Any string value assigned to a subkey is already respected and not treated as a boolean: ``when: complex_variable['any non-empty string']`` is always ``True`` and ``when: not complex_variable['any non-empty string']`` is always ``False``. If you want a string subkey like ``complex_variable['false']`` to be evaluated as a boolean you must use the ``bool`` filter.

Reviewer: Maybe this instead:
```suggestion
The ``conditional_bare_variables`` setting does not affect nested variables. Any string value assigned to a subkey is already respected and not treated as a boolean: in case "complex_variable['subkey']" is a non-empty string, ``when: complex_variable['subkey']`` is always ``True`` and ``when: not complex_variable['subkey']`` is always ``False``. If you want a string subkey like ``complex_variable['subkey']`` (with value `'true'` or `'false'`) to be evaluated as a boolean you must use the ``bool`` filter.
```

This behavior can be changed to be an error or to be ignored by setting the ``ANSIBLE_STRING_CONVERSION_ACTION`` environment variable, or by setting the ``string_conversion_action`` configuration in the ``defaults`` section of ``ansible.cfg``.

Command line facts
------------------

``cmdline`` facts returned in system will be deprecated in favor of ``proc_cmdline``. This change handles special case where Kernel command line parameter contains multiple values with the same key.

Bare variables in conditionals
------------------------------

In Ansible 2.7 and earlier, top-level variables sometimes treated boolean strings as if they were boolean values. This led to inconsistent behavior in conditional tests built on top-level variables defined as strings. Ansible 2.8 began changing this behavior. For example, if you set two conditions like this::

> 

Reviewer 2: better to use a code-block here if we ever translate content.

Reviewer 3: addressed by 708c3f226fefec7486a5022a182269bc6f4aa07d

* In Ansible 2.7 and earlier, both conditions evaluated as ``False`` if ``teardown: 'false'``
* In Ansible 2.8 and later, you have the option of disabling conditional bare variables, so ``when: teardown`` always evaluates as ``True`` and ``when: not teardown`` always evaluates as ``False`` when ``teardown`` is a non-empty string (including ``'true'`` or ``'false'``)

Ultimately, ``when: 'string'`` will always evaluate as ``True`` and ``when: not 'string'`` will always evaluate as ``False``, even if the value of ``'string'`` itself looks like a boolean. For users with playbooks that depend on the old behavior, we added a config setting that preserves it. You can use the ``ANSIBLE_CONDITIONAL_BARE_VARS`` environment variable or ``conditional_bare_variables`` in the ``defaults`` section of ``ansible.cfg`` to select the behavior you want on your control node. The default setting is ``true``, which preserves the old behavior. Set the config value or environment variable to ``false`` to start using the new option. In 2.10 the default will change to ``false``. In 2.12 the old behavior will be deprecated.

> 

Reviewer 2: consider putting the last two sentences in a note so users pay attention to the change coming in later releases.

Reviewer 3: addressed by 708c3f226fefec7486a5022a182269bc6f4aa07d

Updating your playbooks
^^^^^^^^^^^^^^^^^^^^^^^

To prepare your playbooks for the new behavior, you must update your conditional statements so they accept only boolean values. For variables, you can use the ``bool`` filter to evaluate the string ``'false'`` as ``False``::

> 

Reviewer 2: same comment - use code-blocks here and througout.

Reviewer 3: addressed by 708c3f226fefec7486a5022a182269bc6f4aa07d

      - debug:
        when: my_list | length > 0

      - debug:
        when: my_dictionary | length > 0

Do not use the ``bool`` filter with lists or dictionaries. If you use ``bool`` with a list or dict, Ansible will always evaluate it as ``False``.

Double-interpolation
^^^^^^^^^^^^^^^^^^^^

The ``conditional_bare_variables`` setting also affects variables set based on other variables. The old behavior unexpectedly double-interpolated those variables. For example::

    vars:
      double_interpolated: 'bare_variable'
      bare_variable: false

    tasks:
      - debug:
        when: double_interpolated

* In Ansible 2.7 and earlier, ``when: double_interpolated`` evaluated to the value of ``bare_variable``, in this case, ``False``. If the variable ``bare_variable`` is undefined, the conditional fails.
* In Ansible 2.8 and later, with bare variables disabled, Ansible evaluates ``double_interpolated`` as the string ``'bare_variable'``, which is ``True``.

To double-interpolate variable values, use curly braces::

          vars:
            double_interpolated: "{{ other_variable }}"
            other_variable: false

Nested variables
^^^^^^^^^^^^^^^^

> The ``conditional_bare_variables`` setting does not affect nested variables. Subkey strings are already respected and not treated as booleans: ``when: complex_variable['subkey']`` is always ``True`` and ``when: not complex_variable['subkey']`` is always ``False``. If you want a string subkey to be evaluated as a boolean you must use the ``bool`` filter.

Reviewer: I think this should be a bit more explicit. When reading it, it took me quite a while to understand that "subkey strings" did not mean the `'subkey'` string, but that `complex_variable['subkey']` is a string. (Also, if it is the empty string, it will evaluate as `False`.)

Reviewer 3: is this addressed by 708c3f226fefec7486a5022a182269bc6f4aa07d?

Reviewer 3: grr, with 897b805ff71b6d5f222dd089c83932edc95f7c1f to address the empty string case?

> The ``conditional_bare_variables`` setting does not affect nested variables. Subkey strings are already respected and not treated as booleans: ``when: complex_variable['subkey']`` is always ``True`` and ``when: not complex_variable['subkey']`` is always ``False``. If you want a string subkey to be evaluated as a boolean you must use the ``bool`` filter.

Reviewer: I think this should be a bit more explicit. When reading it, it took me quite a while to understand that "subkey strings" did not mean the `'subkey'` string, but that `complex_variable['subkey']` is a string. (Also, if it is the empty string, it will evaluate as `False`.)
